ports/stm32/mboot/fsload.c

```diff
@@ -177,8 +185,13 @@ static int fsload_program_file(bool write_to_flash) {
     if (res != 16) {
         return -MBOOT_ERRNO_DFU_READ_ERROR;
     }
-
-    // TODO validate CRC32
+    // Don't include final 4 bytes; that's the expected crc.
```

Comment:
I think if you do include the final 4 bytes in the calculation, then the output CRC should be expected to be 0 (or 0xffffffff).

Reply:
I'm pretty sure that only works with the XFER form of crc, not the type/polynomial used in dfu.

Reply:
Turns out you're probably right! For ease of testing I used a copy of fwupdate.py https://github.com/micropython/micropython/blob/2ea21abae0358a46c0144f014f2e5891afe03dfd/ports/stm32/mboot/fwupdate.py#L78 I already had with the crc function split out:
``` py
def compute_crc(f):
    """Computes the CRC32 value for the file handle passed in,
    minus the final 4 bytes (which hold the checksum itself)."""
    import ubinascii
    f.seek(0, 2)  # os.SEEK_END
    length = f.tell()
    f.seek(0)

    file_crc = 0
    chunk = 16 * 1024
    buff = bytearray(chunk)
    while length >= chunk + 4:
        read = f.readinto(buff)
        length -= read
        file_crc = ubinascii.crc32(buff, file_crc)

    read = f.readinto(buff)
    final = memoryview(buff)[0 : read - 4]
    file_crc = ubinascii.crc32(final, file_crc)
    file_crc = 0xFFFFFFFF & -file_crc - 1

    expected_crc = ubinascii.hexlify(bytes(reversed(memoryview(buff)[read-4:read]))).decode()
    print("calc: 0x%04x, expect: 0x%s" % (file_crc, expected_crc))

    f.seek(0)
    return file_crc
```

```
>>> f=open('PYBD-SF6-20220117-v1.18.dfu', 'rb')
>>> compute_crc(f)
calc: 0x7b37f517, expect: 0x7b37f517
2067264791
>>>
```

Changing that to include the crc in the calc:
``` py
    read = f.readinto(buff)
<   final = memoryview(buff)[0 : read - 4]
>   final = memoryview(buff)[0 : read]
    file_crc = ubinascii.crc32(final, file_crc)
```

Gives
```
>>> compute_crc(f)
calc: 0x0000, expect: 0x7b37f517
0
>>>
```

Reply:
Yay!  I did check this also using fwupdate.py, which I wanted to push changes for.